stage1/init/common/kvm_mount.go

```diff
@@ -12,260 +12,125 @@
+			}
 		}
 
-		// assertion to make sure that "what" exists (created earlier by PodToSystemdHostMountUnits)
-		diag.Printf("checking required source path: %q", whatFullPath)
-		if _, err := os.Stat(whatFullPath); os.IsNotExist(err) {
-			return fmt.Errorf("bug: missing source for volume %v", vol.Name)
+		readOnly := IsMountReadOnly(vol, app.MountPoints)
+		var source string
+		switch vol.Kind {
+		case "host":
+			source = vol.Source
+		case "empty":
+			source = filepath.Join(common.SharedVolumesPath(root), vol.Name.String())
+		default:
+			return fmt.Errorf(`invalid volume kind %q. Must be one of "host" or "empty"`, vol.Kind)
 		}
-
-		// optionally prepare app directory
-		diag.Printf("optionally preparing destination path: %q", whereFullPath)
-		err := os.MkdirAll(whereFullPath, 0700)
+		absAppRootfs, err := filepath.Abs(common.AppRootfsPath(".", appName))
 		if err != nil {
-			return errwrap.Wrap(fmt.Errorf("failed to prepare dir for mount %v", m.Volume), err)
+			return fmt.Errorf(`could not evaluate absolute path for application rootfs in app: %v`, appName)
 		}
 
-		// install new mount unit for bind mount /mnt/volumeName -> /opt/stage2/{app-id}/rootfs/{{mountPoint.Path}}
-		mu, err := installNewMountUnit(
-			root,      // where put a mount unit
-			whatPath,  // what - stage1 rootfs /mnt/VolumeName
-			wherePath, // where - inside chroot app filesystem
-			"bind",    // fstype
-			"bind",    // options
-			serviceUnitName(appName),
-			unitsDir,
-		)
+		absDestination, err := filepath.Abs(filepath.Join(absAppRootfs, m.Path))
 		if err != nil {
-			return errwrap.Wrap(fmt.Errorf("cannot install new mount unit for app %q", appName.String()), err)
+			return fmt.Errorf(`could not evaluate absolute path for application volume path %q in: %v`, m.Path, appName)
 		}
-
-		// TODO(iaguis) when we update util-linux to 2.27, this code can go
-		// away and we can bind-mount RO with one unit file.
-		// http://ftp.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-ReleaseNotes
-		if IsMountReadOnly(vol, app.MountPoints) {
-			opts := []*unit.UnitOption{
-				unit.NewUnitOption("Unit", "Description", fmt.Sprintf("Remount read-only unit for %s", wherePath)),
-				unit.NewUnitOption("Unit", "DefaultDependencies", "false"),
-				unit.NewUnitOption("Unit", "After", mu),
-				unit.NewUnitOption("Unit", "Wants", mu),
-				unit.NewUnitOption("Service", "ExecStart", fmt.Sprintf("/usr/bin/mount -o remount,ro %s", wherePath)),
-				unit.NewUnitOption("Install", "RequiredBy", mu),
-			}
-
-			remountUnitPath := filepath.Join(root, unitsDir, unit.UnitNamePathEscape(wherePath+"-remount.service"))
-			if err := writeUnit(opts, remountUnitPath); err != nil {
-				return err
-			}
+		if !strings.HasPrefix(absDestination, absAppRootfs) {
+			return fmt.Errorf("path escapes app's root: %v", absDestination)
+		}
+		// TODO: verify if rkt should ensure existance of destination there, or it should be some
+		// outer responsibility to ensure if such path exists
```

Comment:
Actually `PrepareMountpoints` only creates the dir for empty volumes. It looks like `PrepareMountpoints` could be patched to do that.